src/Lecture2.hs

```diff
@@ -207,9 +323,34 @@ The algorithm of merge sort is the following:
 >>> mergeSort [3, 1, 2]
 [1,2,3]
 -}
+
+-- First take.
+-- I think there should be a way to not to use `length` here
+-- but I couldn't find one yet.
+--mergeSort0 :: Ord a => [a] -> [a]
+mergeSort0 :: [Int] -> [Int]
+mergeSort0 [] = []
+mergeSort0 [x] = [x]
+mergeSort0 xs = merge (mergeSort0 left) (mergeSort0 right)
+    where (left, right) = splitAt (length xs `div` 2) xs
+
+-- Maybe something along the line of splitting the list into chunks of (2?)
+-- The thing is I have no idea whether it's more efficient or less,
+-- And it's a different algorithm, although it sorts and merges :)
 mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
+mergeSort [] = []
+mergeSort [x] = [x]
+mergeSort xs = (foldr (merge . sort2) [] . chunkList 2) xs
```

Comment:
That's a clever and correct solution 😅 
Unfortunately, it's not optimal asymptotically. Your solution has complexity `O(n ^ 2)` while optimal merge sort has `O(n log n)`. But the standard `sort` function in Haskell implements an idea similar to yours and improves it a bit to achieve even better performance 🐎

Reply:
Oh, looked at standard `sort`, it's also not comprehensible for me yet :)
One thing I've noticed though, they don't use the optimization you've suggested to `merge`, they don't check if both heads are EQ:
```Haskell
    merge as@(a:as') bs@(b:bs')
      | a `cmp` b == GT = b:merge as  bs'
      | otherwise       = a:merge as' bs
    merge [] bs         = bs
    merge as []         = as
```
Wonder why

Reply:
I also wonder why 🤔 
Maybe it'll be better with such an optimization. But impossible to tell without benchmarks